runtime/tests/checker/transactions_test.go

```diff
@@ -348,6 +365,206 @@ func TestCheckTransactions(t *testing.T) {
+                    self.x = 1
+                }
+              }
+            `,
+			[]error{
+				&sema.InvalidAccessModifierError{},
+			},
+		)
+	})
+}
+
+func TestCheckTransactionRoles(t *testing.T) {
+
+	t.Parallel()
+
+	test := func(t *testing.T, code string, expectedErrors []error) {
+		t.Parallel()
+
+		_, err := ParseAndCheck(t, code)
+
+		errs := RequireCheckerErrors(t, err, len(expectedErrors))
+
+		for i, err := range errs {
+			if !assert.IsType(t, expectedErrors[i], err) {
+				t.Log(err)
+			}
+		}
+	}
+
+	t.Run("empty", func(t *testing.T) {
+		test(
+			t,
+			`
+              transaction {
+
+                  role foo {}
+              }
+            `,
+			nil,
+		)
+	})
+
+	t.Run("field, prepare", func(t *testing.T) {
+		test(
+			t,
+			`
+              transaction {
+
+                  role foo {
+
+                      let bar: Int
+
+                      prepare() {
+                          self.bar = 1
+                      }
+                  }
+              }
+            `,
+			nil,
+		)
+	})
+
+	t.Run("field, missing prepare", func(t *testing.T) {
+		test(
+			t,
+			`
+              transaction {
+
+                  role foo {
+
+                      let bar: Int
+                  }
+              }
+            `,
+			[]error{
+				&sema.MissingPrepareError{},
+			},
+		)
+	})
+
+	t.Run("field, prepare, missing initialization", func(t *testing.T) {
+		test(
+			t,
+			`
+              transaction {
+
+                  role foo {
+
+                      let bar: Int
+
+                      prepare() {}
+                  }
+              }
+            `,
+			[]error{
+				&sema.FieldUninitializedError{},
+			},
+		)
+	})
+
+	t.Run("duplicate", func(t *testing.T) {
+		test(
+			t,
+			`
+              transaction {
+
+                  role foo {}
+
+                  role foo {}
+              }
+            `,
+			[]error{
+				&sema.DuplicateTransactionRoleError{},
+			},
+		)
+	})
+
+	t.Run("field name conflict", func(t *testing.T) {
+		test(
+			t,
+			`
+              transaction {
+
+                  let foo: Int
+
+                  prepare() {
+                      self.foo = 1
+                  }
+
+                  role foo {}
+              }
+            `,
+			[]error{
+				&sema.TransactionRoleWithFieldNameError{},
+			},
+		)
+	})
+
+	t.Run("multiple roles, one field each, execute", func(t *testing.T) {
+		test(
+			t,
+			`
+              transaction {
+
+                  role foo {
+
+                      let bar: Int
+
+                      prepare() {
+                          self.bar = 1
+                      }
+                  }
+
+                  role ying {
+
+                      let yang: String
+
+                      prepare() {
+                          self.yang = "2"
+                      }
+                  }
+
+                  execute {
+                      let bar: Int = self.foo.bar
+                      let yang: String = self.ying.yang
```

Comment:
can we not trivialize chinese philosophy for test cases? also it's yin not ying

Reply:
Huh? I used it because I thought it was nice, I wasn't intentionally trying to trivialize it. 

Changed it in 557080cf6c8b51db54f226980eaab66befb20a14